[v6r12] SRM2Storage, busyfile new return code?; GraphData protect against missing key

andresailer opened the pull request:
a) I am not sure, but now the Castor-SRM return status 16 for SRM_FILE_BUSY. I don't know if this is a new return code for the same issue as before, or a different Issue.

b) I was trying to plot WMSHistory NumberOfJobs grouped by Status, limited to two groups , limited to the last two days and got an exception for a missing key in the parsed_errors

chaen commented:
Hum, this is strange... did you look at the castor specification to see if they don't correspond to different situation ?

andresailer commented:
No idea where I would find that information. Not even sure if this is due to the new dirac version, new externals, lcg bundles, or the castor update for castor public that was done one Wednesday...

chaen commented:
There was a change in the srm-ifce version that changed an 11 to a 22 . it maybe changed that as well... I'll check with Alejandro if that is really the case. Because if it is CASTOR specific, and then the other storage have another error code, then we are doomed :-)

chaen commented:
but wait, which new Dirac version ? I see you commit against v6r12, so you are not speaking about v6r13 ? the one I refer to was only starting at some point in v6r13 if I recall correctly

andresailer commented:
I changed from v6r11 to v6r22 on Wednesday, also changed LCGBundles. Probably also different server externals.

chaen commented:
well I am afraid that without more detailed versions, I can't say very much. But it would still be very interesting to know where would the difference be

zmathe commented on Core/Utilities/Graphs/GraphData.py:
```diff
@@ -401,7 +401,7 @@ def initialize( self ):
       self.keys = self.sortKeys()
 
     self.values = [ self.parsed_data[k] for k in self.keys ]
```
Could you please also add a protection to this line as well?
self.values = [ self.parsed_data[k] if k in self.parsed_data else for k in self.keys ]

zmathe replied:
ops sorry: self.values = [ self.parsed_data[k] if k in self.parsed_data else 0.0 for k in self.keys ]

zmathe commented on Core/Utilities/Graphs/GraphData.py:
```diff
@@ -401,7 +401,7 @@ def initialize( self ):
     self.values = [ self.parsed_data[k] for k in self.keys ]
-    self.errors = [ self.parsed_errors[k] for k in self.keys ]
+    self.errors = [ self.parsed_errors[k] if k in self.parsed_errors else 0.0 for k in self.keys ]
     values_to_sum = [ self.parsed_data[k] for k in self.keys if k != '' ]
```
and also to this line:
values_to_sum = [ self.parsed_data[k] if k in self.parsed_data else 0.0 for k in self.keys if k != '' ]

andresailer commented:
Added the key checks

chaen commented on Core/Utilities/Graphs/GraphData.py:
```diff
@@ -400,9 +400,9 @@ def initialize( self ):
-    self.values = [ self.parsed_data[k] for k in self.keys ]
-    self.errors = [ self.parsed_errors[k] for k in self.keys ]
-    values_to_sum = [ self.parsed_data[k] for k in self.keys if k != '' ]
+    self.values = [ self.parsed_data[k] if k in self.parsed_data else 0.0 for k in self.keys ]
```
self.parsed_data.get(k, 0.0) is maybe clearer isn't it ?

andresailer replied:
Now that you mention it :)

atsareg commented:
Any more details about the SRM return codes ?

andresailer commented:
Not from my side. I can drop this part if people prefer.

atsareg commented:
I will merge this now. @chaen: please try to check if this is Castor specific with Alejandro or whoever appropriate. Would be nice to avoid <if backend == "Castor"...> but may be it is necessary

chaen commented:
I repeat "I am afraid that without more detailed versions, I can't say very much".
But by just looking at the release notes of SRM-IFCE : https://its.cern.ch/jira/browse/DMC-637

andresailer commented:
I think I had moved to the lcgBundles 2015-07-06, after which I noticed that the return code was different. However on the same day there was an upgrade of castorpublic, I have no idea which part is responsible for the error code. By now I moved back to 2015-01-09 on my servers.
